chain/network/src/routing/graph_v2/mod.rs

```diff
@@ -0,0 +1,668 @@
+        if local_edge.edge_type() == EdgeState::Removed {
+            return false;
+        }
+
+        // If the spanning tree doesn't already include the direct edge, add it
+        let mut spanning_tree = distance_vector.edges.clone();
+        if tree_edge.is_none() {
+            debug_assert!(advertised_distances[local_node_id] == -1);
+            spanning_tree.push(local_edge.clone());
+            advertised_distances[local_node_id] = 1;
+        }
+
+        // .min().unwrap() is safe here because the tree is now guaranteed to at least
+        // include the direct edge between the local node and the peer
+        debug_assert!(!spanning_tree.is_empty());
+        let min_nonce = spanning_tree.iter().map(|e| e.nonce()).min().unwrap();
+
+        // Store the tree used to validate the distances.
+        self.edge_cache.update_tree(&distance_vector.root, &spanning_tree);
+        // Store the validated distances
+        self.peer_distances.insert(
+            distance_vector.root.clone(),
+            PeerDistances { distance: advertised_distances, min_nonce },
+        );
+
+        true
+    }
+
+    /// Verifies the given DistanceVector.
+    /// Returns a boolean indicating whether the DistanceVector was valid.
+    /// If applicable, stores the advertised distances for forwarding.
+    fn handle_distance_vector(
+        &mut self,
+        distance_vector: &network_protocol::DistanceVector,
+    ) -> bool {
+        // Basic sanity check; `distance_vector` should come from some other peer
+        if self.config.node_id == distance_vector.root {
+            return false;
+        }
+
+        // Validate the advertised distances against the accompanying spanning tree
+        let validated_distances = self.validate_routing_distances(distance_vector);
+
+        let is_valid = validated_distances.is_some();
+
+        let stored = match validated_distances {
+            Some(distances) => self.store_validated_peer_distances(&distance_vector, distances),
+            None => false,
+        };
+
+        if !stored {
+            // Free ids which may have been allocated to perform validation
+            self.edge_cache.free_unused_ids();
+        }
+
+        return is_valid;
+    }
+
+    /// Handles disconnection of a peer.
+    /// - Updates the state of `local_edges`.
+    /// - Erases the peer's latest spanning tree, if there is one, from `edge_cache`.
+    /// - Erases the advertised distances for the peer.
+    pub(crate) fn remove_direct_peer(&mut self, peer_id: &PeerId) {
+        if let Some(edge) = self.local_edges.get_mut(peer_id) {
+            // TODO(saketh): refactor Edge once the old routing protocol is deprecated
+            if edge.edge_type() != EdgeState::Removed {
+                let (peer0, peer1) = edge.key().clone();
+                // V2 routing protocol doesn't broadcast tombstones; don't bother to sign them
+                *edge = Edge::make_fake_edge(peer0, peer1, edge.nonce() + 1);
```

Comment:
Is fake edge guaranteed to be removed? The assert below is a bit scary.

Reply:
Yeah, again, magic values are a bad idea, better to encode it directly as a bool field in the Edge. Glad to hear it's going away, it would be much safer and cleaner to encode it as a bool field in the edge.